# Leaf Classification---Kaggle competition

## 1. Preparation for training

In [2]:
import numpy as np
import scipy as sp
import pandas as pd
import tensorflow as tf

# Python visulization
import matplotlib.pyplot as plt
import seaborn as sns
import pyecharts as pchart

## Modelling Algorithms
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

%matplotlib inline
%config InlineBackend.figure_format = 'retina'

pd.set_option('display.width', 500)
pd.set_option('display.max_columns', 100)
pd.set_option('display.notebook_repr_html', True)

## 2. Get to knwo the dataset

In [3]:
import os

In [4]:
PATH = os.getcwd()
IMAGES = PATH + "/images"
TRAIN = PATH + "/train.csv"
TEST = PATH + "/test.csv"
SAMPLE = PATH + "sample_submission.csv"

print (PATH)

/Users/charliedang/Documents/Kaggle/LeafClassifier


### 2.1 Basic information of the dataset

In [5]:
## Get the dataset from the file

df = pd.read_csv(TRAIN, sep = ',')

In [6]:
df.head()

id                species   margin1   margin2   margin3   margin4   margin5   margin6   margin7  margin8   margin9  margin10  margin11  margin12  margin13  margin14  margin15  margin16  margin17  margin18  margin19  margin20  margin21  margin22  margin23  margin24  margin25  margin26  margin27  margin28  margin29  margin30  margin31  margin32  margin33  margin34  margin35  margin36  margin37  margin38  margin39  margin40  margin41  margin42  margin43  margin44  margin45  margin46  margin47  \
0   1            Acer_Opalus  0.007812  0.023438  0.023438  0.003906  0.011719  0.009766  0.027344      0.0  0.001953  0.033203  0.013672  0.019531  0.066406  0.000000  0.029297       0.0  0.031250  0.011719  0.000000  0.025391  0.023438  0.001953       0.0  0.015625  0.000000  0.031250  0.000000  0.013672  0.029297  0.015625  0.011719  0.003906  0.025391  0.000000  0.001953  0.011719  0.009766  0.041016  0.037109  0.019531  0.000000  0.009766  0.021484  0.015625  0.007812  0.013672  0.027344   
1   2  Pterocarya_Stenoptera  0.005859  0.000000  0.031250  0.015625  0.025391  0.001953  0.019531      0.0  0.000000  0.007812  0.003906  0.027344  0.023438  0.000000  0.033203       0.0  0.009766  0.009766  0.007812  0.007812  0.019531  0.007812       0.0  0.000000  0.007812  0.027344  0.003906  0.037109  0.007812  0.048828  0.054688  0.027344  0.003906  0.000000  0.000000  0.003906  0.013672  0.033203  0.033203  0.019531  0.031250  0.009766  0.007812  0.031250  0.001953  0.039062  0.029297   
2   3   Quercus_Hartwissiana  0.005859  0.009766  0.019531  0.007812  0.003906  0.005859  0.068359      0.0  0.000000  0.044922  0.007812  0.011719  0.021484  0.001953  0.025391       0.0  0.009766  0.011719  0.007812  0.005859  0.041016  0.001953       0.0  0.015625  0.000000  0.009766  0.001953  0.009766  0.009766  0.015625  0.005859  0.000000  0.017578  0.007812  0.005859  0.009766  0.019531  0.042969  0.021484  0.001953  0.000000  0.005859  0.015625  0.009766  0.011719  0.011719  0.031250   
3   5        Tilia_Tomentosa  0.000000  0.003906  0.023438  0.005859  0.021484  0.019531  0.023438      0.0  0.013672  0.017578  0.001953  0.019531  0.001953  0.003906  0.035156       0.0  0.005859  0.000000  0.001953  0.003906  0.039062  0.009766       0.0  0.009766  0.005859  0.027344  0.001953  0.017578  0.003906  0.021484  0.027344  0.027344  0.000000  0.003906  0.011719  0.003906  0.005859  0.044922  0.056641  0.027344  0.015625  0.007812  0.017578  0.029297  0.001953  0.042969  0.085938   
4   6     Quercus_Variabilis  0.005859  0.003906  0.048828  0.009766  0.013672  0.015625  0.005859      0.0  0.000000  0.005859  0.001953  0.044922  0.041016  0.011719  0.041016       0.0  0.009766  0.015625  0.011719  0.007812  0.021484  0.000000       0.0  0.000000  0.015625  0.009766  0.001953  0.017578  0.041016  0.005859  0.021484  0.013672  0.017578  0.000000  0.005859  0.009766  0.027344  0.083984  0.027344  0.027344  0.001953  0.001953  0.017578  0.007812  0.009766  0.011719  0.080078   

   margin48    ...      texture15  texture16  texture17  texture18  texture19  texture20  texture21  texture22  texture23  texture24  texture25  texture26  texture27  texture28  texture29  texture30  texture31  texture32  texture33  texture34  texture35  texture36  texture37  texture38  texture39  texture40  texture41  texture42  texture43  texture44  texture45  texture46  texture47  texture48  texture49  texture50  texture51  texture52  texture53  texture54  texture55  texture56  texture57  \
0  0.062500    ...            0.0        0.0   0.047852        0.0   0.054688   0.022461        0.0   0.000977   0.018555   0.001953   0.008789   0.015625   0.044922   0.000000   0.037109   0.012695   0.028320   0.000000   0.019531   0.026367   0.005859        0.0   0.004883   0.016602   0.034180   0.056641   0.006836   0.000977   0.022461   0.037109   0.004883   0.021484   0.035156   0.000977   0.004883   0.015625   0.000000        0.0   0.006836   0.037109   0.007812   0.000000   0.002930  

In [7]:
## This is my own defined function-package

import Pdprint as pp

In [8]:
pp.get_info(df)


The Dataset has 194 columns, and 990 observations.
------------------------------------------------


The features of the DataFrame are:  ['id' 'species' 'margin1' 'margin2' 'margin3' 'margin4' 'margin5'
 'margin6' 'margin7' 'margin8' 'margin9' 'margin10' 'margin11' 'margin12'
 'margin13' 'margin14' 'margin15' 'margin16' 'margin17' 'margin18'
 'margin19' 'margin20' 'margin21' 'margin22' 'margin23' 'margin24'
 'margin25' 'margin26' 'margin27' 'margin28' 'margin29' 'margin30'
 'margin31' 'margin32' 'margin33' 'margin34' 'margin35' 'margin36'
 'margin37' 'margin38' 'margin39' 'margin40' 'margin41' 'margin42'
 'margin43' 'margin44' 'margin45' 'margin46' 'margin47' 'margin48'
 'margin49' 'margin50' 'margin51' 'margin52' 'margin53' 'margin54'
 'margin55' 'margin56' 'margin57' 'margin58' 'margin59' 'margin60'
 'margin61' 'margin62' 'margin63' 'margin64' 'shape1' 'shape2' 'shape3'
 'shape4' 'shape5' 'shape6' 'shape7' 'shape8' 'shape9' 'shape10' 'shape11'
 'shape12' 'shape13' 'shape14' 'shape15

In [9]:
pp.list_null(df)

194  columns: 
id : nan 0 ,  int64
species : nan 0 ,  object
margin1 : nan 0 ,  float64
margin2 : nan 0 ,  float64
margin3 : nan 0 ,  float64
margin4 : nan 0 ,  float64
margin5 : nan 0 ,  float64
margin6 : nan 0 ,  float64
margin7 : nan 0 ,  float64
margin8 : nan 0 ,  float64
margin9 : nan 0 ,  float64
margin10 : nan 0 ,  float64
margin11 : nan 0 ,  float64
margin12 : nan 0 ,  float64
margin13 : nan 0 ,  float64
margin14 : nan 0 ,  float64
margin15 : nan 0 ,  float64
margin16 : nan 0 ,  float64
margin17 : nan 0 ,  float64
margin18 : nan 0 ,  float64
margin19 : nan 0 ,  float64
margin20 : nan 0 ,  float64
margin21 : nan 0 ,  float64
margin22 : nan 0 ,  float64
margin23 : nan 0 ,  float64
margin24 : nan 0 ,  float64
margin25 : nan 0 ,  float64
margin26 : nan 0 ,  float64
margin27 : nan 0 ,  float64
margin28 : nan 0 ,  float64
margin29 : nan 0 ,  float64
margin30 : nan 0 ,  float64
margin31 : nan 0 ,  float64
margin32 : nan 0 ,  float64
margin33 : nan 0 ,  float64
margin34 : nan 0 ,  floa

In [13]:
len(df['species'].unique())

99

**Some basic information from the dataset: **

1. The original dataset is made of 194 columns
2. The first two columns are not really useful for training. Id is just a redundant column, but the species column is 
   used to as target or dependant variable
3. There is no nan value in the dataset, so the data cleaning process can be skipped
4. The other features remind us CNN could be used to train the model

In [16]:
## Let's take a look at if every category has same number of observations

df.groupby('species').count()

id  margin1  margin2  margin3  margin4  margin5  margin6  margin7  margin8  margin9  margin10  margin11  margin12  margin13  margin14  margin15  margin16  margin17  margin18  margin19  margin20  margin21  margin22  margin23  margin24  margin25  margin26  margin27  margin28  margin29  margin30  margin31  margin32  margin33  margin34  margin35  margin36  margin37  margin38  margin39  margin40  margin41  margin42  margin43  margin44  margin45  margin46  margin47  \
species                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
Acer_Capillipes               10       10       10       10       10       10       10       10       10       10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10   
Acer_Circinatum               10       10       10       10       10       10       10       10       10       10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10   
Acer_Mono                     10       10       10       10       10       10       10       10       10       10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10   
Acer_Opalus                   10       10       10       10       10       10       10       10       10       10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10   
Acer_Palmatum                 10       10       10       10       10       10       10       10       10       10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10   
Acer_Pictum                   10       10       10       10       10       10       10       10       10       10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10        10   
Acer_Platanoids               10       10       10    

**Conclution: **

Since there are 990 observations in the column, but we have 99 different specises. So that the observations are evenly distributed among these data.